// scryfall-sdk/out/IScry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Color;
(function (Color) {
    Color[Color["W"] = 0] = "W";
    Color[Color["B"] = 1] = "B";
    Color[Color["R"] = 2] = "R";
    Color[Color["U"] = 3] = "U";
    Color[Color["G"] = 4] = "G";
})(Color = exports.Color || (exports.Color = {}));
var Layout;
(function (Layout) {
    Layout[Layout["normal"] = 0] = "normal";
    Layout[Layout["split"] = 1] = "split";
    Layout[Layout["flip"] = 2] = "flip";
    Layout[Layout["transform"] = 3] = "transform";
    Layout[Layout["meld"] = 4] = "meld";
    Layout[Layout["leveler"] = 5] = "leveler";
    Layout[Layout["plane"] = 6] = "plane";
    Layout[Layout["phenomenon"] = 7] = "phenomenon";
    Layout[Layout["scheme"] = 8] = "scheme";
    Layout[Layout["vanguard"] = 9] = "vanguard";
    Layout[Layout["emblem"] = 10] = "emblem";
})(Layout = exports.Layout || (exports.Layout = {}));
var Format;
(function (Format) {
    Format[Format["standard"] = 0] = "standard";
    Format[Format["modern"] = 1] = "modern";
    Format[Format["legacy"] = 2] = "legacy";
    Format[Format["vintage"] = 3] = "vintage";
    Format[Format["commander"] = 4] = "commander";
    Format[Format["pauper"] = 5] = "pauper";
    Format[Format["frontier"] = 6] = "frontier";
    Format[Format["penny"] = 7] = "penny";
    Format[Format["duel"] = 8] = "duel";
    Format[Format["1v1"] = 9] = "1v1";
    Format[Format["future"] = 10] = "future";
})(Format = exports.Format || (exports.Format = {}));
var Legality;
(function (Legality) {
    Legality[Legality["legal"] = 0] = "legal";
    Legality[Legality["not_legal"] = 1] = "not_legal";
    Legality[Legality["restricted"] = 2] = "restricted";
    Legality[Legality["banned"] = 3] = "banned";
})(Legality = exports.Legality || (exports.Legality = {}));
var Rarity;
(function (Rarity) {
    Rarity[Rarity["common"] = 0] = "common";
    Rarity[Rarity["uncommon"] = 1] = "uncommon";
    Rarity[Rarity["rare"] = 2] = "rare";
    Rarity[Rarity["mythic"] = 3] = "mythic";
})(Rarity = exports.Rarity || (exports.Rarity = {}));
exports.Frame = {
    "1993": 0,
    "2003": 1,
    "2015": 2,
    future: 3,
    0: "1993",
    1: "2003",
    2: "2015",
    3: "future",
};
var Border;
(function (Border) {
    Border[Border["black"] = 0] = "black";
    Border[Border["white"] = 1] = "white";
    Border[Border["silver"] = 2] = "silver";
    Border[Border["gold"] = 3] = "gold";
})(Border = exports.Border || (exports.Border = {}));
var SetType;
(function (SetType) {
    SetType[SetType["core"] = 0] = "core";
    SetType[SetType["expansion"] = 1] = "expansion";
    SetType[SetType["masters"] = 2] = "masters";
    SetType[SetType["masterpiece"] = 3] = "masterpiece";
    SetType[SetType["from_the_vault"] = 4] = "from_the_vault";
    SetType[SetType["premium_deck"] = 5] = "premium_deck";
    SetType[SetType["duel_deck"] = 6] = "duel_deck";
    SetType[SetType["commander"] = 7] = "commander";
    SetType[SetType["planechase"] = 8] = "planechase";
    SetType[SetType["conspiracy"] = 9] = "conspiracy";
    SetType[SetType["archenemy"] = 10] = "archenemy";
    SetType[SetType["vanguard"] = 11] = "vanguard";
    SetType[SetType["funny"] = 12] = "funny";
    SetType[SetType["starter"] = 13] = "starter";
    SetType[SetType["box"] = 14] = "box";
    SetType[SetType["promo"] = 15] = "promo";
    SetType[SetType["token"] = 16] = "token";
})(SetType = exports.SetType || (exports.SetType = {}));
